From 154c56b78a764154c99ab0b6856936b951e447d5 Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Mon, 16 Feb 2026 09:44:47 +0000
Subject: [PATCH 05/13] TelephonyManager: add prop to force SIM slot count

Some devices report an incorrect number of SIM slots, causing dual-SIM
functionality to be unavailable. Add persist.sys.phh.sim_slots property
to override the reported SIM slot count in getMultiSimConfiguration(),
getActiveModemCount(), and getSupportedModemCount().

Setting persist.sys.phh.sim_slots=2 forces DSDS mode, =3 forces TSTS,
=0 or =1 forces single-SIM. When unset (-1), the default behavior is
preserved.
---
 .../android/telephony/TelephonyManager.java   | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/telephony/java/android/telephony/TelephonyManager.java b/telephony/java/android/telephony/TelephonyManager.java
index 773558bd8..1bb2f9bb8 100644
--- a/telephony/java/android/telephony/TelephonyManager.java
+++ b/telephony/java/android/telephony/TelephonyManager.java
@@ -646,6 +646,15 @@ public class TelephonyManager {
     /** @hide */
     @UnsupportedAppUsage
     public MultiSimVariants getMultiSimConfiguration() {
+        int simSlots = SystemProperties.getInt("persist.sys.phh.sim_slots", -1);
+        if (simSlots == 0 || simSlots == 1) {
+            return MultiSimVariants.UNKNOWN;
+        } else if (simSlots == 2) {
+            return MultiSimVariants.DSDS;
+        } else if (simSlots == 3) {
+            return MultiSimVariants.TSTS;
+        }
+
         String mSimConfig =
                 TelephonyProperties.multi_sim_config().orElse("");
         if (mSimConfig.equals("dsds")) {
@@ -687,6 +696,11 @@ public class TelephonyManager {
      * </ul>
      */
     public int getActiveModemCount() {
+        int simSlots = SystemProperties.getInt("persist.sys.phh.sim_slots", -1);
+        if (simSlots >= 0) {
+            return simSlots;
+        }
+
         int modemCount = 1;
         switch (getMultiSimConfiguration()) {
             case UNKNOWN:
@@ -715,6 +729,11 @@ public class TelephonyManager {
      * {@link #getActiveModemCount} returns 1 while this API returns 2.
      */
     public int getSupportedModemCount() {
+        int simSlots = SystemProperties.getInt("persist.sys.phh.sim_slots", -1);
+        if (simSlots >= 0) {
+            return simSlots;
+        }
+
         return TelephonyProperties.max_active_modems().orElse(getActiveModemCount());
     }
 
-- 
2.53.0

